### DISCUSSION TITLE
Action Streams

### DISCUSSION THREAD

**Discussion by aretrace** [withastro/roadmap#1015] [Proposal]
# Summary

Have actions be able to return streams.

# Background & Motivation

I like [streams](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html) 😁

# Goals

- You can only return serialized JSON, this is too limiting for todays applications. Enabling different modes follows the philosophy of actions; currently we have two input (`accept`) modes, 'json' (implicit) and 'form' (explicit). Having mirror parameters for action outputs (e.g., `{ remit: 'stream' }`) would be very useful.

# [Example](https://elysiajs.com/patterns/stream.html) 🐺

```js
defineAction({
  accept: '...',
  remit: 'stream',
  handler: async function* (input, ctx) {
    ...
    yield ...
    ...
  }
})
```
✨✨✨✨✨✨✨✨
&nbsp;some [inspiration](https://github.com/TanStack/router/blob/main/docs/framework/react/start/server-functions.md#accepting-parameters)
✨✨✨✨✨✨✨✨

**Comment by delucis** [comment]
Interesting. Could you share an example use case or two? When might you need to stream the result back?

**Reply by aretrace** [reply]
In essence, any scenario where a series of data is incrementally supplied, possibly without a defined endpoint. 

Examples include:
- Real-time updates
- Progress indicators
- Instant messaging
- Monitoring and logging
- IoT and sensor data
- Notifications

Right now a popular use case would be relaying LLM response streams. 
The participants of https://github.com/facebook/react-native/issues/27741 could also explain why it's beneficial to stream a return.

**Comment by aretrace** [comment]
### 👇 Underlying Implementation Ideas 👇
>  i.e., *http 1.1 chunked transfer encoding, server-sent events, `application/octet-stream`, etc...*
